components/feature/session/src/main/java/mozilla/components/feature/session/view/VerticalSwipeRefreshLayout.kt

```diff
@@ -0,0 +1,180 @@
+ *
+ * Following situations will not trigger pull to refresh:
+ * - a scroll happening more on the horizontal axis
+ * - a scale in/out gesture
+ * - a quick scale gesture
+ *
+ * To control responding to scrolls and showing the pull to refresh throbber or not
+ * use the [View.isEnabled] property.
+ */
+class VerticalSwipeRefreshLayout @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null
+) : SwipeRefreshLayout(context, attrs) {
+    @VisibleForTesting
+    internal var isQuickScaleInProgress = false
+    @VisibleForTesting
+    internal var quickScaleEvents: Triple<MotionEvent?, MotionEvent?, MotionEvent?> = Triple(null, null, null)
+    private var previousX = 0f
+    private var previousY = 0f
+    private val doubleTapTimeout = ViewConfiguration.getDoubleTapTimeout()
+    private val doubleTapSlop = ViewConfiguration.get(context).scaledDoubleTapSlop
+    private val doubleTapSlopSquare = doubleTapSlop * doubleTapSlop
+
+    @Suppress("ComplexMethod", "ReturnCount")
+    override fun onInterceptTouchEvent(event: MotionEvent): Boolean {
+        // Setting "isEnabled = false" is recommended for users of this ViewGroup
+        // who who are not interested in the pull to refresh functionality
+        // Setting this easily avoids executing code unneededsly before the check for "canChildScrollUp".
+        if (!isEnabled) {
+            return false
+        }
+
+        // Layman's scale gesture (with two fingers) detector.
+        // Allows for quick, serial inference as opposed to using ScaleGestureDetector
+        // which uses callbacks and would be hard to synchronize in the little time we have.
+        if (event.pointerCount > 1) {
+            return false
+        }
+
+        val eventAction = event.action
+
+        // Cleanup if the gesture has been aborted or quick scale just ended/
+        if (MotionEvent.ACTION_CANCEL == eventAction ||
+            (MotionEvent.ACTION_UP == eventAction && isQuickScaleInProgress)) {
+
+            forgetQuickScaleEvents()
+            return callSuperOnInterceptTouchEvent(event)
+        }
+
+        // Disable pull to refresh if quick scale is in progress.
+        maybeAddDoubleTapEvent(event)
+        if (isQuickScaleInProgress(quickScaleEvents)) {
+            isQuickScaleInProgress = true
+            return false
+        }
+
+        // Disable pull to refresh if the move was more on the X axis.
+        if (MotionEvent.ACTION_DOWN == eventAction) {
+            previousX = event.x
+            previousY = event.y
+        } else if (MotionEvent.ACTION_MOVE == eventAction) {
+            val xDistance = abs(event.x - previousX)
+            val yDistance = abs(event.y - previousY)
+            previousX = event.x
+            previousY = event.y
+            if (xDistance > yDistance) {
+                return false
+            }
+        }
+
+        return callSuperOnInterceptTouchEvent(event)
```

Comment:
same here

Reply:
All the extra call* are used in tests to check if super.* was called.
Did not know of another way other than this proxies.